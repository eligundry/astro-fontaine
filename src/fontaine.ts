import {
  generateFontFace,
  generateFallbackName,
  getMetricsForFamily,
  readMetrics,
} from 'fontaine'
import * as csstree from 'css-tree'
import fs from 'node:fs/promises'
import path from 'node:path'
import { Readable } from 'node:stream'
import { createWriteStream } from 'node:fs'
import { finished } from 'node:stream/promises'
import { pathToFileURL } from 'node:url'
import crypto from 'node:crypto'

interface Payload {
  href: string
  family: string
  fallbacks: string[]
  fontDirectory?: string
}

interface CustomFont {
  family: string
  fallbacks?: string[]
  src?: string
}

interface FontFromStylesheet extends CustomFont {
  src: string
}

function hashHref(href: string) {
  return `${crypto.createHash('md5').update(href).digest('hex')}.css`
}

async function getCachedCSS(
  href: string,
  fontDirectory: string
): Promise<string | undefined> {
  const key = hashHref(href)
  const cachedCssPath = path.join(fontDirectory, key)
  const fileExists = await checkIfFileExists(cachedCssPath)

  if (fileExists) {
    return fs.readFile(cachedCssPath, 'utf-8')
  }

  return undefined
}

async function cacheCSS(href: string, fontDirectory: string, css: string) {
  const key = hashHref(href)
  const cachedCssPath = path.join(fontDirectory, key)
  await fs.writeFile(cachedCssPath, css)
}

async function checkIfFileExists(file: string) {
  try {
    await fs.access(file, fs.constants.R_OK)
    return true
  } catch (e) {
    return false
  }
}

export async function generateCSS({
  href,
  fallbacks,
  fontDirectory = './public/astro-fontaine',
}: Payload): Promise<string> {
  const cachedCSS = await getCachedCSS(href, fontDirectory)

  if (cachedCSS) {
    return cachedCSS
  }

  // Download the stylesheet
  const stylesheet = await fetch(href).then((resp) => {
    if (!resp.ok) {
      throw new Error(`Could not fetch the following stylesheet URL: ${href}`)
    }

    return resp.text()
  })

  // Extract all of the font URLs
  const ast = csstree.parse(stylesheet)
  const fontFaceDeclarations: Array<FontFromStylesheet | CustomFont> = csstree
    .findAll(ast, (node, _item, _list) => {
      return node.type === 'Atrule' && node.name === 'font-face'
    })
    .map((ruleAst): FontFromStylesheet | null => {
      if (
        ruleAst.type !== 'Atrule' ||
        !ruleAst.block ||
        !ruleAst.block.children
      ) {
        return null
      }

      let family: string = ''
      let src: string = ''

      for (const node of ruleAst.block.children) {
        if (node.type === 'Declaration' && node.value.type === 'Value') {
          if (node.property === 'src') {
            // @ts-ignore
            src = node.value.children?.head?.data?.value?.trim()
          } else if (node.property === 'font-family') {
            // @ts-ignore
            family = node.value.children?.head?.data?.value?.trim()
          }
        }
      }

      if (!src || !family) {
        return null
      }

      return {
        family,
        src,
        fallbacks,
      }
    })
    .filter(Boolean)

  // Download all of the fonts to the app's public directory.
  // Fonts load the fastest when severed from the site's domain, as the
  // browser no longer does cross site caching and it'll save on
  // redundent request logic.
  await Promise.all(
    fontFaceDeclarations.map(async (fontFace) => {
      if (!fontFace.src) {
        return
      }

      const parsedURL = new URL(fontFace.src)
      const pathToSaveFont = path.join(
        fontDirectory,
        parsedURL.host,
        ...parsedURL.pathname.split('/')
      )
      const fileExists = await checkIfFileExists(pathToSaveFont)

      // If we already have the font, skip downloading it.
      if (fileExists) {
        return true
      }

      await fs.mkdir(path.dirname(pathToSaveFont), { recursive: true })

      const font = await fetch(fontFace.src).then((resp) => {
        if (!resp.ok) {
          throw new Error(`Could not download font ${fontFace.src}`)
        }

        if (!resp.body) {
          throw new Error(`Downloaded font is empty ${fontFace.src}`)
        }

        return resp.body as ReadableStream<Uint8Array>
      })

      const stream = createWriteStream(pathToSaveFont)
      // @ts-ignore
      await finished(Readable.fromWeb(font).pipe(stream))
    })
  )

  // Generate the font fallback metrics
  const fallbackFontsCSS = await Promise.all(
    fontFaceDeclarations
      // We only need to generate the fallback once per family, as style and
      // weight are not used in the fallback.
      .filter(
        (value, index, self) =>
          index === self.findIndex((f) => f.family === value.family)
      )
      .map(async (fontFace) => {
        let metrics = await getMetricsForFamily(fontFace.family)

        if (!metrics && fontFace.src) {
          const parsedURL = new URL(fontFace.src)
          const file = path.join(
            fontDirectory,
            parsedURL.host,
            ...parsedURL.pathname.split('/')
          )
          metrics = await readMetrics(pathToFileURL(file))
        }

        if (!metrics) {
          console.warn('Could not finde metrics for font', fontFace)
          return ''
        }

        return generateFontFace(metrics, {
          name: generateFallbackName(fontFace.family),
          fallbacks: fontFace.fallbacks,
        })
      })
  )

  // Update the AST with the new font URLs
  // In order to save on HTTP overhead, we serve the fonts from the same
  // hosts as the site.
  csstree.walk(ast, (node) => {
    if (
      node.type === 'Url' &&
      fontFaceDeclarations.find(({ src }) => src === node.value)
    ) {
      node.value = node.value.replace('https://', '/astro-fontaine/')
    }
  })

  const generatedCSS = csstree.generate(ast) + '\n' + fallbackFontsCSS

  await cacheCSS(href, fontDirectory, generatedCSS)

  return generatedCSS
}
