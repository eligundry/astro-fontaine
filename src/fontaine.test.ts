import { describe, it, expect, beforeEach, vi } from 'vitest'
import fsSync from 'node:fs'
import fs from 'node:fs/promises'
import path from 'node:path'
import { generateCSS } from './fontaine'

describe('generateCSS', () => {
  vi.mock('node:fs/promises')

  beforeEach(() => {
    fetchMock.resetMocks()
    vi.restoreAllMocks()
  })

  it('should return cached file if it exists', async () => {
    vi.mocked(fs.access).mockResolvedValueOnce(void 0)
    vi.mocked(fs.readFile).mockResolvedValueOnce('body { color: red }')

    expect(
      await generateCSS({
        href: 'https://fonts.googleapis.com/css2?family=Inter',
        family: 'Inter',
        fallbacks: [],
      })
    ).toBe('body { color: red }')
  })

  it('should successfully parse and transform a @font-face stylesheet', async () => {
    fetchMock.mockResponseOnce(() => {
      return {
        body: `@font-face {
  font-family: 'Inter';
  font-style: normal;
  font-weight: 400;
  src: url(https://fonts.gstatic.com/s/inter/v13/UcCO3FwrK3iLTeHuS_fvQtMwCp50KnMw2boKoduKmMEVuLyfAZ9hiJ-Ek-_EeA.woff2) format('woff2');
}`,
      }
    })

    expect(
      await generateCSS({
        href: 'https://fonts.googleapis.com/css2?family=Inter',
        family: 'Inter',
        fallbacks: ['Helvetica', 'Arial'],
      })
    ).toMatchSnapshot()
  })

  it('should remove `display=fallback` from Google Fonts URLs', async () => {
    fetchMock.mockResponseOnce((req) => {
      const parsed = new URL(req.url)
      expect(parsed.searchParams.get('display')).toBeFalsy()

      return {
        body: `@font-face {
  font-family: 'Inter';
  font-style: normal;
  font-weight: 400;
  src: url(https://fonts.gstatic.com/s/inter/v13/UcCO3FwrK3iLTeHuS_fvQtMwCp50KnMw2boKoduKmMEVuLyfAZ9hiJ-Ek-_EeA.woff2) format('woff2');
}`,
      }
    })

    expect(
      await generateCSS({
        href: 'https://fonts.googleapis.com/css2?family=Inter&display=fallback',
        family: 'Inter',
        fallbacks: ['Helvetica', 'Arial'],
      })
    ).toMatchSnapshot()
  })

  it('should handle Typekit URLs', async () => {
    const stylesheetURL = 'https://use.typekit.net/foobar.css'
    const font = fsSync.readFileSync(
      path.join(__dirname, 'fixtures', 'font.woff2'),
      { encoding: 'binary' }
    )

    vi.mocked(fs.mkdir).mockResolvedValue(undefined)
    vi.mocked(fs.writeFile).mockResolvedValue(undefined)

    fetchMock.mockOnceIf(stylesheetURL, () => ({
      body: `
@font-face {
font-family:"Foobar";
src:url("https://use.typekit.net/fonts/xxxxxx/xxxxxxxxxxxxxxxxxxxxxxxx/xx/l?primer=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx&fvd=n4&v=3") format("woff2"),url("https://use.typekit.net/fonts/xxxxx/xxxxxxxxxxxxxxxxxxxxxxxx/xx/d?primer=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx&fvd=n4&v=3") format("woff"),url("https://use.typekit.net/fonts/bbbbbb/xxxxxxxxxxxxxxxxxxxxxxxx/xx/x?primer=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx&fvd=n4&v=3") format("opentype");
font-display:auto;font-style:normal;font-weight:400;font-stretch:normal;
}
.tk-foobar { font-family: "Foobar",sans-serif; }
      `,
    }))

    fetchMock.mockIf(/^https:\/\/use.typekit.net\/fonts/, async () => {
      return {
        body: font.toString(),
      }
    })

    expect(
      await generateCSS({
        href: stylesheetURL,
        family: 'Foobar',
        fallbacks: ['Helvetica', 'Arial'],
      })
    ).toMatchSnapshot()
  })
})
