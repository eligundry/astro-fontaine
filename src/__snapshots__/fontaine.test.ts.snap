// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateCSS > should remove \`display=fallback\` from Google Fonts URLs 1`] = `
"@font-face{font-family:\\"Inter\\";font-style:normal;font-weight:400;src:url(/astro-fontaine/fonts.gstatic.com/s/inter/v13/UcCO3FwrK3iLTeHuS_fvQtMwCp50KnMw2boKoduKmMEVuLyfAZ9hiJ-Ek-_EeA.woff2)format(\\"woff2\\")}
@font-face {
  font-family: \\"Inter fallback\\";
  src: local(\\"Helvetica\\");
  size-adjust: 100%;
  ascent-override: 96.875%;
  descent-override: 24.1477%;
  line-gap-override: 0%;
}

@font-face {
  font-family: \\"Inter fallback\\";
  src: local(\\"Arial\\");
  size-adjust: 100%;
  ascent-override: 96.875%;
  descent-override: 24.1477%;
  line-gap-override: 0%;
}"
`;

exports[`generateCSS > should successfully parse and transform a @font-face stylesheet 1`] = `
"@font-face{font-family:\\"Inter\\";font-style:normal;font-weight:400;src:url(/astro-fontaine/fonts.gstatic.com/s/inter/v13/UcCO3FwrK3iLTeHuS_fvQtMwCp50KnMw2boKoduKmMEVuLyfAZ9hiJ-Ek-_EeA.woff2)format(\\"woff2\\")}
@font-face {
  font-family: \\"Inter fallback\\";
  src: local(\\"Helvetica\\");
  size-adjust: 100%;
  ascent-override: 96.875%;
  descent-override: 24.1477%;
  line-gap-override: 0%;
}

@font-face {
  font-family: \\"Inter fallback\\";
  src: local(\\"Arial\\");
  size-adjust: 100%;
  ascent-override: 96.875%;
  descent-override: 24.1477%;
  line-gap-override: 0%;
}"
`;
